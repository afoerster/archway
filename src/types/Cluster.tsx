import { Dropdown, Icon, Menu } from 'antd';
import * as Color from 'color';
import * as React from 'react';
import Colors from '../components/Colors';

export interface WebLocation {
    host: string;
    port: number;
}

export interface Statusable {
  status: string;
}

export class Status<T extends Statusable> {
    public statusable: T;

    constructor(statusable: T) {
    this.statusable = statusable;
  }

    public statusColor = (): Color => {
    switch (this.statusable.status) {
      case 'GOOD_HEALTH':
        return Colors.Green;
      case 'CONCERNING_HEALTH':
        return Colors.Orange;
      case 'BAD_HEALTH':
        return Colors.Red;
      default:
        return Colors.Gray;
    }
  }

    public statusText = () => {
    switch (this.statusable.status) {
      case 'GOOD_HEALTH':
        return '"good"';
      case 'CONCERNING_HEALTH':
        return '"concerning"';
      case 'BAD_HEALTH':
        return '"bad"';
      default:
        return 'unknown';
    }
  }

    public glowColorText = () =>
    `0 0 5px 2px ${this.statusColor().hsl().string()}`
}

export interface HueService extends Statusable {
  load_balancer: WebLocation[];
}

/* tslint:disable:no-empty-interface */
export interface HiveService extends Statusable {

}

export interface YarnService extends Statusable {
    resource_manager: WebLocation[];
    node_manager: WebLocation[];
}

export abstract class ServiceLinks<T extends Statusable> {
    public service: T;

    public links: JSX.Element[];

    constructor(service: T) {
    this.service = service;
  }
}

export class HueServiceLinks extends ServiceLinks<HueService> {
    public links: JSX.Element[] =
    this.service && this.service.load_balancer.map((location: WebLocation) => (
      <a target="_blank" rel="noreferrer noopener" href={`https://${location.host}:${location.port}`}>Hue UI</a>
    ));
}

export class HiveServiceLinks extends ServiceLinks<HiveService> {
  public links: JSX.Element[] = [];
}

export class YarnServiceLinks extends ServiceLinks<YarnService> {
    public resourceManagerLinks: JSX.Element[] =
    this.service && this.service.resource_manager.map((location) => (
      <Menu.Item key={location.host}>
        <a target="_blank" rel="noreferrer noopener" href={`https://${location.host}:${location.port}`}>
          {location.host}
        </a>
      </Menu.Item>
    ));

    public nodeManagerLinks: JSX.Element[] =
    this.service && this.service.node_manager.map((location) => (
      <Menu.Item key={location.host}>
        <a target="_blank" rel="noreferrer noopener" href={`https://${location.host}:${location.port}`}>
          {location.host}
        </a>
      </Menu.Item>
    ));

    public links: JSX.Element[] = [
    (
      <Dropdown overlay={<Menu>{this.nodeManagerLinks}</Menu>}>
        <a href="#" className="ant-dropdown-link">
          Node Manager UI <Icon type="down" />
        </a>
      </Dropdown>
    ),
    (
      <Dropdown overlay={<Menu>{this.resourceManagerLinks}</Menu>}>
        <a href="#" className="ant-dropdown-link">
          Resource Manager UI <Icon type="down" />
        </a>
      </Dropdown>
    ),
  ];
}

export interface Cluster extends Statusable {
    name: string;
    cm_url: string;
    services: {
      hive: HiveService
      hue: HueService
      yarn: YarnService,
  };
}
