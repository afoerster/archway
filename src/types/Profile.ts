export interface Profile {
  name: string;
  username: string;
  permissions: {
    risk_management: boolean
    platform_operations: boolean,
  };
}
