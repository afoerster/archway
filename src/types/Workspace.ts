export interface YarnApplication {
  id: string;
  name: string;
  start_time?: Date;
}

export interface ResourcePoolsInfo {
  loading: boolean;
  error?: string;
  data: PoolInfo[];
}

export interface PoolInfo {
  name: string;
  applications: YarnApplication[];
}

export interface NamespaceInfo {
  name: string;
  tables: HiveTable[];
}

export interface HiveTable {
  name: string;
}

export interface MemberRemoveStatus {
  loading?: boolean;
  success?: boolean;
  error?: boolean;
}

export interface Member {
  username: string;
  name: string;
  removeStatus?: MemberRemoveStatus;
}

export interface Application {
  id: number;
  name: string;
  consumer_group: string;
  group: SecurityGroup;
}

export interface TopicGrant {
  id: number;
  group: SecurityGroup;
  actions: string;
  topic_access?: Date;
}

export interface KafkaTopic {
  id: number;
  name: string;
  partitions: number;
  replication_factor: number;
  managing_role: TopicGrant;
  readonly_role: TopicGrant;
}

export interface ResourcePool {
  id: number;
  pool_name: string;
  max_cores: number;
  max_memory_in_gb: number;
}

export interface SecurityGroup {
  common_name: string;
  distinguished_name: string;
  sentry_role: string;
  group_created?: Date;
  role_created?: Date;
  role_associated?: Date;
}

export interface Compliance {
  phi_data: boolean;
  pci_data: boolean;
  pii_data: boolean;
}

export interface DatabaseGrant {
  group: SecurityGroup;
  location_access?: Date;
  database_access?: Date;
}

export interface HiveAllocation {
  id: number;
  name: string;
  location: string;
  size_in_gb: number;
  consumed_in_gb: number;
  managing_group: DatabaseGrant;
  readonly_group?: DatabaseGrant;
}

export interface ApprovalStatus {
  loading?: boolean;
  success?: boolean;
  error?: string;
}

export interface ApprovalItem {
  approver?: string;
  approval_time?: string;
  status?: ApprovalStatus;
}

export interface Approvals {
  infra?: ApprovalItem;
  risk?: ApprovalItem;
}

export interface Workspace {
  id: number;
  name: string;
  summary: string;
  description: string;
  behavior: string;
  requested_date: Date;
  requester: string;
  single_user: boolean;
  status?: string;
  compliance: Compliance;
  approvals?: Approvals;
  data: HiveAllocation[];
  processing: ResourcePool[];
  topics: KafkaTopic[];
  applications: Application[];
}

export interface UserSuggestion {
  display: string;
  common_name: string;
}

export interface UserSuggestions {
  filter: string;
  users: UserSuggestion[];
  groups: UserSuggestion[];
}
