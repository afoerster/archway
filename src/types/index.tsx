export interface Login {
    email: String
    password: String
}

export interface Auth {
    token: String
    error: String
    loading?: Boolean
    loggingIn: Boolean
}

export interface StoreState {
    auth: Auth
}