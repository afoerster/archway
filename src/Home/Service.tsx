import * as React from 'react';
import {Card, Avatar} from 'antd';
import {Statusable, Status, ServiceLinks} from '../types/Cluster';

export interface Props {
  name: String
  status: Status<Statusable>
  links: ServiceLinks<Statusable>
  index: number
}

const ServiceDisplay = ({name, status, links, index}: Props) => (
  <Card
    style={{ flex: 1, marginLeft: index == 0 ? 0 : 25 }}
    actions={links.links}>
    <Card.Meta
      title={name}
      description={`${name}'s status is currently ${status.statusText()}`}
      avatar={
        <Avatar
          size="small"
          style={{ boxShadow: status.glowColorText(), backgroundColor: status.statusColor().string() }} />
      } />
  </Card>
);

export default ServiceDisplay;