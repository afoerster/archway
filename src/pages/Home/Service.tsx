import { Avatar, Card } from 'antd';
import * as React from 'react';
import { ServiceLinks, Status, Statusable } from '../../types/Cluster';

export interface Props {
    name: string;
    status: Status<Statusable>;
    links: ServiceLinks<Statusable>;
    index: number;
}

const ServiceDisplay = ({ name, status, links, index }: Props) => (
  <Card
    style={{ flex: 1, marginLeft: index === 0 ? 0 : 25 }}
    actions={links.links}>
    <Card.Meta
      title={name}
      description={`${name}'s status is currently ${status.statusText()}`}
      avatar={
        <Avatar
          size="small"
          style={{ boxShadow: status.glowColorText(), backgroundColor: status.statusColor().string() }} />
      } />
  </Card>
);

export default ServiceDisplay;
