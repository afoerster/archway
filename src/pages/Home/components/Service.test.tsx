import * as React from 'react';
import { shallow } from 'enzyme';

import Service from './Service';
import { HiveServiceLinks, Status, Statusable } from '../../../types/Cluster';

describe('Service', () => {
  it('renders correctly', () => {
    const status = new Status<Statusable>({
      status: '',
    });
    const links = new HiveServiceLinks({
      status: '',
    });
    const wrapper = shallow(
      <Service
        name=""
        status={status}
        links={links}
        index={0}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
