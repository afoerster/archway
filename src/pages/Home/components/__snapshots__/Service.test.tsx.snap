// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ServiceDisplay
    index={0}
    links={
      HiveServiceLinks {
        "links": Array [],
        "service": Object {
          "status": "",
        },
      }
    }
    name=""
    status={
      Status {
        "glowColorText": [Function],
        "statusColor": [Function],
        "statusText": [Function],
        "statusable": Object {
          "status": "",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "actions": Array [],
      "children": <Unknown
        avatar={
          <Avatar
            prefixCls="ant-avatar"
            shape="circle"
            size="small"
            style={
              Object {
                "backgroundColor": "rgb(187, 187, 187)",
                "boxShadow": "0 0 5px 2px hsl(0, 0%, 73.3%)",
              }
            }
          />
        }
        description="'s status is currently unknown"
        title=""
      />,
      "style": Object {
        "flex": 1,
        "marginLeft": 0,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "avatar": <Avatar
          prefixCls="ant-avatar"
          shape="circle"
          size="small"
          style={
            Object {
              "backgroundColor": "rgb(187, 187, 187)",
              "boxShadow": "0 0 5px 2px hsl(0, 0%, 73.3%)",
            }
          }
        />,
        "description": "'s status is currently unknown",
        "title": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "actions": Array [],
        "children": <Unknown
          avatar={
            <Avatar
              prefixCls="ant-avatar"
              shape="circle"
              size="small"
              style={
                Object {
                  "backgroundColor": "rgb(187, 187, 187)",
                  "boxShadow": "0 0 5px 2px hsl(0, 0%, 73.3%)",
                }
              }
            />
          }
          description="'s status is currently unknown"
          title=""
        />,
        "style": Object {
          "flex": 1,
          "marginLeft": 0,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "avatar": <Avatar
            prefixCls="ant-avatar"
            shape="circle"
            size="small"
            style={
              Object {
                "backgroundColor": "rgb(187, 187, 187)",
                "boxShadow": "0 0 5px 2px hsl(0, 0%, 73.3%)",
              }
            }
          />,
          "description": "'s status is currently unknown",
          "title": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
